internal/github/github.go

```diff
@@ -164,138 +180,113 @@ func (c CIStatus) String() string {
 	}
 }
 
-func (c GHClient) getCIChecksResult(ctx context.Context, owner, repoName, ref string) (CIStatus, error) {
-	listOptions := github.ListOptions{
-		PerPage: 100,
+func (c GHClient) GetCIStatus(ctx context.Context, owner, repoName, ref string) (CIStatus, error) {
+	var query struct {
+		Repository struct {
+			Object struct {
+				Commit struct {
+					StatusCheckRollup struct {
+						State    string `graphql:"state"`
+						Contexts struct {
+							Nodes []struct {
+								Typename string `graphql:"__typename"`
+								CheckRun struct {
+									Name       string `graphql:"name"`
+									Conclusion string `graphql:"conclusion"`
+								} `graphql:"... on CheckRun"`
+								StatusContext struct {
+									Context string `graphql:"context"`
+									State   string `graphql:"state"`
+								} `graphql:"... on StatusContext"`
+							} `graphql:"nodes"`
+							PageInfo struct {
+								EndCursor   *string `graphql:"endCursor"`
+								HasNextPage bool    `graphql:"hasNextPage"`
+							} `graphql:"pageInfo"`
+						} `graphql:"contexts(first: 100, after: $cursor)"`
+					} `graphql:"statusCheckRollup"`
+				} `graphql:"... on Commit"`
+			} `graphql:"object(expression: $ref)"`
+		} `graphql:"repository(owner: $owner, name: $repository)"`
 	}
 
-	opt := &github.ListCheckRunsOptions{
-		ListOptions: listOptions,
-		Filter:      github.String("latest"),
+	vars := map[string]interface{}{
+		"owner":      graphql.String(owner),
+		"repository": graphql.String(repoName),
+		"ref":        graphql.String(ref),
+		"cursor":     (*graphql.String)(nil),
 	}
 
-	status := CIStatusPending
+	hasChecks := false
+	hasStatuses := false
+	retried := false
 
 	for {
-		runs, resp, err := c.client.Checks.ListCheckRunsForRef(ctx, owner, repoName, ref, opt)
+		err := c.graphQLClient.Query(ctx, &query, vars)
 		if err != nil {
 			return CIStatusUnknown, fmt.Errorf("failed to query GitHub: %w", err)
 		}
 
-		if runs.GetTotal() == 0 {
-			log.Info("No checks found, waiting for a bit to see if one appears")
-			time.Sleep(c.pendingRecheckTime)
-
-			runs, resp, err = c.client.Checks.ListCheckRunsForRef(ctx, owner, repoName, ref, opt)
-			if err != nil {
-				return CIStatusUnknown, fmt.Errorf("failed to query GitHub: %w", err)
-			}
-
-			if runs.GetTotal() == 0 {
-				log.Debug("No checks found after waiting, assuming success")
-				return CIStatusPassed, nil
-			}
+		// return early if all checks and statuses are successful. no need to evaluate individual nodes in the response.
+		if strings.ToLower(query.Repository.Object.Commit.StatusCheckRollup.State) == "success" {
```

Comment:
Suggest extracting `query.Repository.Object.Commit.StatusCheckRollup` to a variable as you access it a few times & this would make the code a bit easier to read.